### PR TITLE
Add a PollDurationMs metric to AbstractKafkaBasedConnectorTask

### PR DESCRIPTION
**Pull request by somandal**
Adding a PollDurationMs metric to track how long KafkaConsumer poll() blocks for Kafka based connectors. This can be disabled via a config.

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Review by jzakaryan — DISMISSED** [review]
lgtm

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java [hunk 1]
```diff
@@ -478,6 +489,10 @@ protected void trackEventsProcessedProgress(int recordCount) {
       }
       _consumerMetrics.updateNumPolls(1);
       _consumerMetrics.updateEventCountsPerPoll(records.count());
+      if (_enablePollDurationMsMetric) {
+        _dynamicMetricsManager.createOrUpdateHistogram(_metricsPrefix, _datastreamName, POLL_DURATION_MS,
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:493]: Do you think refactoring `createKafkaBasedConnectorTaskMetrics` with additional boolean which decides the creation of histogram for poll duration will make more sense? Then in the future, if we decide to enable it for other tasks, we can just set the boolean to true?

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorConfig.java [hunk 2]
```diff
@@ -31,6 +31,7 @@
   public static final String CONFIG_RETRY_SLEEP_DURATION_MILLIS = "retrySleepDurationMs";
   public static final String CONFIG_PAUSE_PARTITION_ON_ERROR = "pausePartitionOnError";
   public static final String CONFIG_PAUSE_ERROR_PARTITION_DURATION_MILLIS = "pauseErrorPartitionDurationMs";
+  public static final String CONFIG_ENABLE_POLL_DURATION_MS_METRIC = "enablePollDurationMsMetric";
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorConfig.java:34]: nit: Looking at the naming convention used in this file, MS is named as MILLIS

**Review by vmaheshw — APPROVED** [review]

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java (hunk 1 above)
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:493]: So the problem is that even if I use the boolean in `createKafkaBasedConnectorTaskMetrics` to create the metric, I'll have to update the metric here, and update it under the same boolean. So I thought why not just create the metric lazily. Plus, this metric lives in `AbstractKafkaBasedConnectorTask`, which means that any connectors that extend this class will get this metric already. Each connector can also decide they don't want this metric, and disable it via config.

Do let me know if I misunderstood your suggestion, and if you're okay with me not adding code to register this metric in `createKafkaBasedConnectorTaskMetrics`.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorConfig.java (hunk 2 above)
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorConfig.java]: done

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java [hunk 3]
```diff
@@ -99,6 +111,10 @@ public void deregisterMetrics() {
     DYNAMIC_METRICS_MANAGER.unregisterMetric(_className, _key, NUM_AUTO_PAUSED_PARTITIONS_ON_INFLIGHT_MESSAGES);
     DYNAMIC_METRICS_MANAGER.unregisterMetric(_className, _key, NUM_AUTO_PAUSED_PARTITIONS_WAITING_FOR_DEST_TOPIC);
     DYNAMIC_METRICS_MANAGER.unregisterMetric(_className, _key, NUM_TOPICS);
+
+    if (_enablePollDurationMillisMetric) {
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java:115]: nit:Can you use _pollDurationMs directly and make _enablePollDurationMillisMetric local?

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java [hunk 4]
```diff
@@ -52,7 +56,11 @@
   private final AtomicLong _numAutoPausedPartitionsAwaitingDestTopic = new AtomicLong(0);
   private final AtomicLong _numTopics = new AtomicLong(0);
 
-  KafkaBasedConnectorTaskMetrics(String className, String metricsKey, Logger errorLogger) {
+  private final boolean _enablePollDurationMillisMetric;
+  private final Histogram _pollDurationMs;
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java:60]: nit: please change the name from _pollDurationMs to _pollDurationMsMetric. the name is confusing.

**Review by vmaheshw — APPROVED** [review]

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java (hunk 4 above)
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java]: done

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java (hunk 3 above)
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java]: done

**Review by vmaheshw — APPROVED** [review]
LGTM

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java (hunk 1 above)
Comment (vmaheshw) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java]: Resolving after offline discussion.

**Review by jzakaryan — APPROVED** [review]

**Pull request merged**